pkg/certs/certs.go

```diff
@@ -26,177 +30,317 @@ import (
-			return nil, err
-		}
+	manager = &Manager{
+		certificates: map[pair]*tls.Certificate{},
+		defaultCert: pair{
+			KeyFile:  keyFile,
+			CertFile: certFile,
+		},
+		loadX509KeyPair: loadX509KeyPair,
+		events:          make(chan notify.EventInfo, 1),
+		ctx:             ctx,
 	}
-
-	return c, nil
-}
-
-func checkSymlink(file string) (bool, error) {
-	st, err := os.Lstat(file)
-	if err != nil {
-		return false, err
+	if err := manager.AddCertificate(certFile, keyFile); err != nil {
+		return nil, err
 	}
-	return st.Mode()&os.ModeSymlink == os.ModeSymlink, nil
+	go manager.watchFileEvents()
+	return manager, nil
 }
 
-// watchSymlinks reloads symlinked files since fsnotify cannot watch
-// on symbolic links.
-func (c *Certs) watchSymlinks() (err error) {
-	cert, err := c.loadCert(c.certFile, c.keyFile)
+// AddCertificate adds the TLS certificate in certFile resp. keyFile
+// to the Manager.
+//
+// If there is already a certificate with the same base name it will be
+// replaced by the newly added one.
+func (m *Manager) AddCertificate(certFile, keyFile string) (err error) {
+	certFile, err = filepath.Abs(certFile)
 	if err != nil {
 		return err
 	}
-	c.Lock()
-	c.cert = &cert
-	c.Unlock()
-	go func() {
-		for {
-			select {
-			case <-c.e:
-				// Once stopped exits this routine.
-				return
-			case <-time.After(24 * time.Hour):
-				cert, cerr := c.loadCert(c.certFile, c.keyFile)
-				if cerr != nil {
-					continue
-				}
-				c.Lock()
-				c.cert = &cert
-				c.Unlock()
-			}
-		}
-	}()
-	return nil
-}
-
-// watch starts watching for changes to the certificate
-// and key files. On any change the certificate and key
-// are reloaded. If there is an issue the loading will fail
-// and the old (if any) certificates and keys will continue
-// to be used.
-func (c *Certs) watch() (err error) {
-	defer func() {
-		if err != nil {
-			// Stop any watches previously setup after an error.
-			notify.Stop(c.e)
-		}
-	}()
-
-	// Windows doesn't allow for watching file changes but instead allows
-	// for directory changes only, while we can still watch for changes
-	// on files on other platforms. Watch parent directory on all platforms
-	// for simplicity.
-	if err = notify.Watch(filepath.Dir(c.certFile), c.e, eventWrite...); err != nil {
+	keyFile, err = filepath.Abs(keyFile)
+	if err != nil {
 		return err
 	}
-	if err = notify.Watch(filepath.Dir(c.keyFile), c.e, eventWrite...); err != nil {
+	certFileIsLink, err := isSymlink(certFile)
+	if err != nil {
 		return err
 	}
-	cert, err := c.loadCert(c.certFile, c.keyFile)
+	keyFileIsLink, err := isSymlink(keyFile)
 	if err != nil {
 		return err
 	}
-	c.Lock()
-	c.cert = &cert
-	c.Unlock()
+	if certFileIsLink && !keyFileIsLink {
+		return fmt.Errorf("certs: '%s' is a symlink but '%s' is a regular file", certFile, keyFile)
+	}
+	if keyFileIsLink && !certFileIsLink {
+		return fmt.Errorf("certs: '%s' is a symlink but '%s' is a regular file", keyFile, certFile)
+	}
```

Comment:
we have observed with @harshavardhana that kubernetes sometimes mount files as symlinks inside containers so this might be an issue, i'll test it.

Reply:
Hm, are you saying K8S may mount file `A` as regular file and may mount file `B` as symlink?!
If that's the case I'm not sure whether we can reliably reload certificates. For example, if K8S mounts the private key as regular file and the certificate as symlink we cannot observe that the certificate gets replaced by K8S / operator...

Reply:
when its symLink previous code was to reload every 24hrs

Reply:
Current code does that too. However, it assumes that both the private key and certificate are symlinks or regular files. This is about one file being a a symlink the other being a regular one.

Reply:
@aead both files will be mounted as symlinks, ie:

```
/tmp/certs # ls -la
total 0
drwxrwxrwt    3 root     root           140 Aug 25 18:18 .
drwxrwxrwt    1 root     root            19 Aug 25 18:18 ..
drwxr-xr-x    3 root     root           100 Aug 25 18:18 ..2020_08_25_18_18_26.626755347
lrwxrwxrwx    1 root     root            31 Aug 25 18:18 ..data -> ..2020_08_25_18_18_26.626755347
lrwxrwxrwx    1 root     root            10 Aug 25 18:18 CAs -> ..data/CAs
lrwxrwxrwx    1 root     root            18 Aug 25 18:18 private.key -> ..data/private.key
lrwxrwxrwx    1 root     root            17 Aug 25 18:18 public.crt -> ..data/public.crt
```

Reply:
Yes we should test this thoroughly..

Reply:
if both files are symlink then there should be no problem correct?

Reply:
Yes, PR should not change the existing behavior - it reloads symlinks every 24h. Testing never hurts, though...